pkg/hooks/webhook.go

```diff
@@ -63,27 +73,70 @@ func NewWebhookExecutor(
 		hookType,
 		client,
 		url)
+	if err != nil {
+		return nil, err
+	}
+	return NewWebhookExecutorWithCustomHttpClient(webhook, hookType, client)
+}
+
+func NewWebhookExecutorWithCustomHttpClient(
+	webhook *v1alpha1.Webhook,
+	hookType common.HookType,
+	httpClient HttpClientInterface,
+) (*WebhookExecutor, error) {
+	if webhook == nil {
+		return nil, nil
+	}
+	url, err := webhookURL(webhook)
+	if err != nil {
+		return nil, err
+	}
+	if err != nil {
+		logging.Logger.Info(err.Error())
+	}
+
+	var eTagCache *etag_cache.Cache = nil
+	if webhook.Etag != nil && webhook.Etag.Enabled != nil && *webhook.Etag.Enabled {
+		eTagCache = etag_cache.NewCache(webhook.Etag.CacheTimeoutSeconds, webhook.Etag.CacheTimeoutSeconds)
+	}
+
 	if err != nil {
 		return nil, err
 	}
 	return &WebhookExecutor{
-		client:   client,
-		url:      url,
-		hookType: hookType.String(),
+		client:    httpClient,
+		url:       url,
+		hookType:  hookType.String(),
+		etagCache: eTagCache,
 	}, nil
 }
 
-func (w *WebhookExecutor) Call(request interface{}, response interface{}) error {
+func (w *WebhookExecutor) Call(request WebhookRequest, response interface{}) error {
 	// Encode request.
 	reqBody, err := k8sjson.Marshal(request)
 	if err != nil {
 		return fmt.Errorf("can't marshal request: %w", err)
 	}
+
+	cacheKey := request.GetCacheKey()
+	cacheEnabled := w.etagCache != nil
+	cacheEntry, cacheEntryExists := w.etagCache.Get(cacheKey)
+	eTagValue := ""
+
+	req, err := http.NewRequest("POST", w.url, bytes.NewReader(reqBody))
+	if err != nil {
+		return err
+	}
+	if cacheEnabled && cacheEntryExists {
+		req.Header.Set("If-None-Match", cacheEntry.Etag)
+		eTagValue = cacheEntry.Etag
+	}
 	if logging.Logger.V(6).Enabled() {
 		rawRequest := json.RawMessage(reqBody)
-		logging.Logger.Info("Webhook request", "type", w.hookType, "url", w.url, "body", rawRequest)
+		logging.Logger.Info("Webhook request", "type", w.hookType, "url", w.url, "etag", eTagValue, "body", rawRequest)
```

Comment:
should make level 6 instead of info `Logger.V(6)`

Reply:
fixed